util/licence-checker.py

```diff
@@ -0,0 +1,357 @@
+# checked.
+COMMENT_CHARS = [
+    # Hardware Files
+    ([".svh", ".sv", ".sv.tpl"], SLASH_SLASH),  # SystemVerilog
+    # Hardware Build Systems
+    ([".tcl"], HASH),  # tcl
+    ([".core", ".core.tpl"], DifferentFirstLineCommentStyle("#", "CAPI=2")),  # FuseSoC Core Files
+    (["Makefile", ".mk"], HASH),  # Makefiles
+    ([".ys"], HASH),  # Yosys script
+    ([".waiver"], HASH),  # Wavier files
+    ([".vlt"], SLASH_SLASH),  # More wavier files
+    ([".vbl"], HASH),  # Verible configuration
+    ([".el", ".el.tpl"], SLASH_SLASH), # Exclusion List
+    ([".do"], SLASH_SLASH),
+    ([".sdc"], HASH),
+    ([".f"], None),  # Dependency Files
+    # Software Files
+    ([".c", ".c.tpl", ".h", ".h.tpl", ".cc", ".cpp"], SLASH_SLASH),  # C,C++
+    ([".S"], SLASH_SLASH),  # Assembly
+    ([".ld"], BlockCommentStyle("/*", "*/")),  # Linker Scripts
+    ([".rs"], SLASH_SLASH),  # Rust
+    # Software Build Systems
+    (["meson.build", "toolchain.txt", "meson_options.txt"], HASH),  # Meson
+    # General Tooling
+    ([".py"], HASH),  # Python
+    ([".sh"], HASH),  # Shell Scripts
+    (["Dockerfile"], HASH),  # Dockerfiles
+    # Configuration
+    ([".hjson", ".hjson.tpl"], SLASH_SLASH),  # hjson
+    ([".yml", ".yaml"], HASH),  # YAML
+    ([".toml"], HASH),  # TOML
+    (["-requirements.txt"], HASH),  # Apt + Python Requirements
+    (["redirector.conf"], HASH),  # nginx config
+    # Documentation
+    ([".md", ".md.tpl", ".html"], None),  # Markdown and HTML
+    ([".css"], BlockCommentStyle("/*", "*/")),  # CSS
+    ([".scss"], SLASH_SLASH),  # SCSS
+    # Unsupported: Explicitly Skip These
+    ([".png", ".svg", ".ico"], None), # Image formats
+    ([], None),
+]
+
+
+def detect_comment_char(filename):
+    for (suffixes, commentstyle) in COMMENT_CHARS:
+        for suffix in suffixes:
+            if filename.endswith(suffix):
+                return commentstyle
+
+    return None
+
+
+def git_find_repo_toplevel():
+    git_output = subprocess.check_output(
+        ['git', 'rev-parse', '--show-toplevel'])
+    return Path(git_output.decode().strip()).resolve()
+
+
+def git_find_all_file_paths(top_level, search_paths):
+    git_output = subprocess.check_output(
+        ["git", "-C", str(top_level), "ls-files", "-z", "--", *search_paths])
+    git_output = git_output.rstrip(b"\0")
+    for path in git_output.split(b"\0"):
+        yield Path(top_level, path.decode())
+
+class ResultsTracker(object):
+    """Helper for tracking results"""
+
+    def __init__(self, base_dir):
+        self.base_dir = base_dir
+
+    passed_count = 0
+    failed_count = 0
+    excluded_count = 0
+    skipped_count = 0
+
+    failing_paths = set()
+
+    @property
+    def total_count(self):
+        return self.passed_count + self.failed_count + self.skipped_count + self.excluded_count
+
+    def passed(self, path, line_no, reason):
+        rel_path = path.relative_to(self.base_dir)
+        logging.debug("%s:%d PASSED: %s", str(rel_path), line_no, reason)
```

Comment:
Don't use "logging" directly (which gives you the root logger). Set up a logger and use it (typically a global `log = logging.getLogger(__name__)`, and then `log.debug()`)

Reply:
You cannot use .basicConfig with a logger set up like this though, so I'm keeping it this way, as this is effectively how every other single-file util in OT sets up logging.

Reply:
You still use `logging.basicConfig()` to set up initial logging, but then use a non-root logger.

I'd prefer if we don't add new scripts which use the root logger for logging, but not blocking this one.